### ISSUE TITLE
Flair Model gives different results on same data

### ISSUE THREAD

**Issue by sawankumar94** [flairNLP/flair#1404]
I trained a flair model with (flair + bert english) embedding on 17000 observations to classify the text into multiple classes. 

When i use the trained "best model" to classify the set of texts, i get different results different time i.e on same text or string it gives different classes after (restarting the kernel -Jupyter Notebook Python). Why?

**Comment by yosipk** [comment]
Have you made sure that you're running the model in evaluation mode and not in the training mode? After loading the model, call `model.eval()` to set the model in evaluation mode.

**Comment by sawankumar94** [comment]
Hi ! Thanks for your reply. I tried your suggestion. I called model.eval() after loading and before prediction. But, still no success.

**Comment by djstrong** [comment]
I am not sure if BERT is deterministic.

**Comment by djstrong** [comment]
@sawankumar94 Could you show your code for loading model and prediction?

**Comment by sawankumar94** [comment]
The code is as follows:

import flair
from flair.models import TextClassifier
from flair.data import Sentence

classifier = TextClassifier.load('bert-best-model.pt')
def func1(new_sentences, model):
    sentences = [Sentence(txt) for txt in new_sentences]
    model.eval()
    model.predict(sentences)
    d= {}
    count = 0
    for sentence in sentences: 
        d[new_sentences[count]] = str(sentence.labels[0]).split()[0]     
        count+=1
    return(d)

pred = func(new_sentences, model = classifier)
print(pred)

**Comment by sawankumar94** [comment]
`import flair
from flair.models import TextClassifier
from flair.data import Sentence

classifier = TextClassifier.load('bert-best-model.pt')
def func1(new_sentences, model):
    sentences = [Sentence(txt) for txt in new_sentences]
    model.eval()
    model.predict(sentences)
    d= {}
    count = 0
    for sentence in sentences: 
        d[new_sentences[count]] = str(sentence.labels[0]).split()[0]     
        count+=1
    return(d)

pred = func(new_sentences, model = classifier)
print(pred)`

**Comment by sawankumar94** [comment]
[check.docx](https://github.com/flairNLP/flair/files/4148236/check.docx)

**Comment by sawankumar94** [comment]
The code is there in the attached file.

**Comment by djstrong** [comment]
The code looks ok.

**Comment by sawankumar94** [comment]
@alanakbik  Please help!

**Comment by sawankumar94** [comment]
This model has dropout too.

**Comment by djstrong** [comment]
Create and share minimal working example.

**Comment by alanakbik** [comment]
@sawankumar94 which BERT embeddings are you using? Could you share the embedding setup of your model training? 

Could you also train a model without BERT embeddings to see if the error still occurs?

**Comment by sawankumar94** [comment]
Below is the code which i used for training the model. I used default Bert Embedding. 

@djstrong  @alanakbik  Please have a look.

`import numpy as np
import pandas as pd
from flair.models import TextClassifier
from flair.data import Sentence


OOS = pd.read_csv('20200201_oversample_pre-holdout_final.txt')
del OOS['Unnamed: 0']
OOS = OOS.rename(columns={"Three_class_sentiment":"label", "sentence":"text"})
OOS = OOS[["label","text"]]


from flair.data_fetcher import NLPTaskDataFetcher
from flair.embeddings import WordEmbeddings, FlairEmbeddings, DocumentLSTMEmbeddings, DocumentRNNEmbeddings, BertEmbeddings
from flair.models import TextClassifier
from flair.trainers import ModelTrainer
from pathlib import Path


from flair.data import Corpus
from flair.datasets import CSVClassificationCorpus
from flair.datasets import ClassificationCorpus


corpus = NLPTaskDataFetcher.load_classification_corpus(Path('./'), test_file='test.csv', dev_file='dev.csv', train_file='train.csv')



label_dict=corpus.make_label_dictionary()

bert_embedding = BertEmbeddings()


max_tokens = 256
corpus._train = [x for x in corpus.train if len(x) < max_tokens]
corpus._dev =   [x for x in corpus.dev if len(x) < max_tokens]
corpus._test =  [x for x in corpus.test if len(x) < max_tokens]

word_embeddings = [bert_embedding, #WordEmbeddings('glove'),
                   FlairEmbeddings('news-forward-fast'), 
                   FlairEmbeddings('news-backward-fast')]

document_embeddings =DocumentRNNEmbeddings(word_embeddings,
                                           hidden_size= 512, 
                                           reproject_words=True, 
                                           reproject_words_dimension=256,
                                           rnn_type='LSTM',
                                           bidirectional= True,
                                           dropout = 0.2
                                          )


classifier = TextClassifier(document_embeddings, label_dictionary = label_dict, multi_label= False)

trainer = ModelTrainer(classifier, corpus)


trainer.train('Bert', 
              learning_rate = 0.05,
              mini_batch_size  = 8,
              eval_mini_batch_size  = 8,
              max_epochs = 16,
              anneal_factor = 0.5,
              patience = 2,
              min_learning_rate = 0.0001,
              train_with_dev = False,
              monitor_train= True,
              monitor_test = True,
              embeddings_storage_mode = 'cpu' ,  #'cpu'
              checkpoint = True,
              save_final_model = True,
              #anneal_with_restarts = False,
              shuffle = True,
              param_selection_mode = False,
              #num_workers = 2 #2 ,
              
              #sampler=None,
              #use_amp = False,
              #amp_opt_level = 'O1'
             )       
`
I used the downloaded 'best-model.pt' for prediction.

I used below code to load the model and predict on list of texts.

`import flair
from flair.models import TextClassifier
from flair.data import Sentence


classifier = TextClassifier.load('bert-best-model.pt')
def document_classification(new_sentences, model):
                            
    #new_sentences is the list of texts to be classified
    sentences = [Sentence(txt) for txt in (new_sentences)]

    model.eval()
    model.predict(sentences)
    output_sentence = []
    count = 0

    for sentence in sentences: 
        if str(sentence.labels[0]).split()[0] == "NE" :
            output_sentence.append(new_sentences[count])     
        count+=1
    
    return(output_sentence)

predictions =  document_classification(new_sentences, classifier)  
print(predictions)                                           `

**Comment by djstrong** [comment]
Could you share the model and a sentence that gives different results for each run?

**Comment by sawankumar94** [comment]
Below is the code which i used for training the model. I used default Bert Embedding.
@djstrong @alanakbik Please have a look.

```
import` numpy as np
import pandas as pd
from flair.models import TextClassifier
from flair.data import Sentence

OOS = pd.read_csv('20200201_oversample_pre-holdout_final.txt')
del OOS['Unnamed: 0']
OOS = OOS.rename(columns={"Three_class_sentiment":"label", "sentence":"text"})
OOS = OOS[["label","text"]]

from flair.data_fetcher import NLPTaskDataFetcher
from flair.embeddings import WordEmbeddings, FlairEmbeddings, DocumentLSTMEmbeddings, DocumentRNNEmbeddings, BertEmbeddings
from flair.models import TextClassifier
from flair.trainers import ModelTrainer
from pathlib import Path

from flair.data import Corpus
from flair.datasets import CSVClassificationCorpus
from flair.datasets import ClassificationCorpus

corpus = NLPTaskDataFetcher.load_classification_corpus(Path('./'), test_file='test.csv', dev_file='dev.csv', train_file='train.csv')

label_dict=corpus.make_label_dictionary()

bert_embedding = BertEmbeddings()

max_tokens = 256
corpus._train = [x for x in corpus.train if len(x) < max_tokens]
corpus._dev = [x for x in corpus.dev if len(x) < max_tokens]
corpus._test = [x for x in corpus.test if len(x) < max_tokens]

word_embeddings = [bert_embedding, #WordEmbeddings('glove'),
FlairEmbeddings('news-forward-fast'),
FlairEmbeddings('news-backward-fast')]

document_embeddings =DocumentRNNEmbeddings(word_embeddings,
hidden_size= 512,
reproject_words=True,
reproject_words_dimension=256,
rnn_type='LSTM',
bidirectional= True,
dropout = 0.2
)

classifier = TextClassifier(document_embeddings, label_dictionary = label_dict, multi_label= False)

trainer = ModelTrainer(classifier, corpus)

trainer.train('Bert',
learning_rate = 0.05,
mini_batch_size = 8,
eval_mini_batch_size = 8,
max_epochs = 16,
anneal_factor = 0.5,
patience = 2,
min_learning_rate = 0.0001,
train_with_dev = False,
monitor_train= True,
monitor_test = True,
embeddings_storage_mode = 'cpu' , #'cpu'
checkpoint = True,
save_final_model = True,
shuffle = True,
param_selection_mode = False)       
```
I used the downloaded 'best-model.pt' for prediction.

I used below code to load the model and predict on list of texts.

```
import flair
from flair.models import TextClassifier
from flair.data import Sentence

classifier = TextClassifier.load('bert-best-model.pt')
def document_classification(new_sentences, model):

#new_sentences is the list of texts to be classified
sentences = [Sentence(txt) for txt in (new_sentences)]

model.eval()
model.predict(sentences)
output_sentence = []
count = 0

for sentence in sentences: 
    if str(sentence.labels[0]).split()[0] == "NE" :
        output_sentence.append(new_sentences[count])     
    count+=1

return(output_sentence)
predictions = document_classification(new_sentences, classifier)
print(predictions)
 ```

**Comment by djstrong** [comment]
I have run your code and the score is the same.
